core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -676,14 +728,36 @@ private OptionalInt getNumberOfTasks(StageId stageId)
 
         private void addDynamicFilters(Map<DynamicFilterId, List<Domain>> newDynamicFilters)
         {
+            SetMultimap<Consumer<Map<DynamicFilterId, Domain>>, DynamicFilterId> completedConsumers = HashMultimap.create();
             newDynamicFilters.forEach((filter, domain) -> {
                 if (taskDynamicFilters.remove(filter) == null) {
                     // filter has been collected concurrently
                     return;
                 }
-                dynamicFilterSummaries.put(filter, union(domain));
+                Domain dynamicFilterDomain = union(domain);
+                dynamicFilterSummaries.put(filter, dynamicFilterDomain);
                 Optional.ofNullable(lazyDynamicFilters.get(filter)).ifPresent(future -> future.set(null));
                 dynamicFilterCollectionTime.put(filter, System.nanoTime());
+                List<Consumer<Map<DynamicFilterId, Domain>>> consumers;
+                synchronized (dynamicFilterConsumers) {
+                    consumers = dynamicFilterConsumers.remove(filter);
+                    if (consumers == null) {
+                        return;
+                    }
+                }
+                // dynamic filter updates are batched up per-consumer to reduce number of callbacks
+                consumers.forEach(consumer -> completedConsumers.put(consumer, filter));
+            });
+            completedConsumers.asMap().forEach((consumer, dynamicFilterIds) -> {
+                Map<DynamicFilterId, Domain> collectedDomain = dynamicFilterIds.stream()
+                        .map(filterId -> {
+                            Domain domain = requireNonNull(
+                                    dynamicFilterSummaries.get(filterId),
+                                    () -> format("Missing dynamic filter %s", filterId));
```

Comment:
this cannot really happen since we just collected filter, `requireNonNull` is sufficient